cmds/netcat/netcat.go

```diff
@@ -1,37 +1,182 @@
-// Copyright 2012 the u-root Authors. All rights reserved
+// Copyright 2016 the u-root Authors. All rights reserved
 // Use of this source code is governed by a BSD-style
 // license that can be found in the LICENSE file.
 
-/*
-netcat connects to a place and sends data to it and from it.
-*/
-
 package main
 
 import (
-	"fmt"
+	"flag"
 	"io"
+	"log"
 	"net"
 	"os"
 )
 
-func main() {
-	var c net.Conn
-	var err error
-	if len(os.Args) < 2 {
-		os.Exit(1)
+// Ready to handle full-size UDP datagram or TCP segment in one step
+const (
+	bufferLimit = 2<<16 - 1
+)
+
+type progress struct {
+	ra     net.Addr
+	rBytes uint64
+	wBytes uint64
+}
+
+// Launch two read-write goroutines and waits for signal from them
+func transferStreams(con net.Conn) {
+	c1 := readAndWrite(con, os.Stdout)
+	c2 := readAndWrite(os.Stdin, con)
+	select {
+	case status := <-c1:
+		log.Printf("Remote connection is closed: %+v\n", status)
+	case status := <-c2:
+		log.Printf("Local program is terminated: %+v\n", status)
 	}
+}
 
-	if c, err = net.Dial("tcp", os.Args[1]); err != nil {
-		fmt.Printf("%v\n", err)
-		os.Exit(1)
+// Launch receive goroutine first, wait for address from it (if needed),
+// launch send goroutine then.
+func transferPackets(con net.Conn) {
+	c1 := readAndWrite(con, os.Stdout)
+	// If connection hasn't got remote address then wait for it from receiver goroutine
+	ra := con.RemoteAddr()
+	if ra == nil {
+		progress := <-c1
+		ra = progress.ra
+		log.Println("Connect from", ra)
+	}
+	c2 := readAndWriteToAddr(os.Stdin, con, ra)
+	select {
+	case progress := <-c1:
+		log.Printf("Remote connection is closed: %+v\n", progress)
+	case progress := <-c2:
+		log.Printf("Local program is terminated: %+v\n", progress)
 	}
+}
+
+// Read from Reader and write to Writer until EOF.
+func readAndWrite(r io.Reader, w io.Writer) <-chan progress {
+	return readAndWriteToAddr(r, w, nil)
+}
+
+// Read from Reader and write to Writer until EOF.
+// ra is an address to whom packets must be sent in UDP listen mode.
+func readAndWriteToAddr(r io.Reader, w io.Writer, ra net.Addr) <-chan progress {
+	buf := make([]byte, bufferLimit)
+	c := make(chan progress)
+	rBytes, wBytes := uint64(0), uint64(0)
 	go func() {
-		if _, err := io.Copy(c, os.Stdin); err != nil {
-			fmt.Printf("%v", err)
+		defer func() {
```

Comment:
having such a massive closure is inefficient and doesn't make the code very readable (it indents it one level deeper than necessary). it would be easier if you changed this function to accept a channel and then pass the progress channel to it, fully controlling it from the caller. in this way if, later, you want to pass a 'cancel' channel it can be done without significant refactoring.

```
c := make(chan progress)
go readAndWriteToAddr(..., c)
for z := <-c, etc
```

also, i'm in favour of simplifying the function names: readAndWrite -> IO, readAndWriteToAddr can become IOAddr, or something similarly short.